### *Capstone Project for Udacity Data Engineering Nanodegree*

# Book Reviews Data Lake

___

# Project Summary

This project goals to build a **Book Reviews Data Lake** containing data on books from across the world, along with Amazon reviews about them. For this purpose, we are going to use data sources provided by [Book Depository](https://www.bookdepository.com/) and [Amazon](https://www.amazon.com/).

The project follows these steps:
* Step 1: Scope the Project and Gather Data
* Step 2: Explore and Assess the Data
* Step 3: Define the Data Model
* Step 4: Run ETL to Model the Data
* Step 5: Complete Project Write Up

## Installs and imports

In [ ]:
!pip install pandas

In [1]:
import os
import configparser
import pandas as pd
import logging
from pyspark.sql import SparkSession
import pyspark.sql.types as t
import pyspark.sql.functions as f

pd.set_option('display.max_columns', 500)

## Configuration file

Open the configuration file `btl.cfg` and set your AWS credentials (`AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY`). The AWS User needs to have at least the **AmazonS3FullAccess** permission policy attached.

Then, set the INPUT parameters `AMAZON_REVIEWS` and `BOOK_DEPOSITORY` with the path from where those datasets will be loaded. 

*To ease the review of this Capstone Project, the files of the Book Depository Dataset were already downloaded from Kaggle and uploaded to a S3 Bucket, and the parameters `AMAZON_REVIEWS` and `BOOK_DEPOSITORY` are already set to the paths of the S3 buckets containing the datasets.*

At last, set the `OUTPUT_DATA` parameter with the path where the Data Lake will be hosted.

```bash
[AWS]
AWS_ACCESS_KEY_ID=
AWS_SECRET_ACCESS_KEY=

[INPUT]
AMAZON_REVIEWS=
BOOK_DEPOSITORY=

[OUTPUT]
DATA_LAKE_PATH=
```

Reading the configuration file `bdl.cfg`:

In [2]:
config = configparser.ConfigParser()
config.read('bdl.cfg')

os.environ['AWS_ACCESS_KEY_ID'] = config['AWS']['AWS_ACCESS_KEY_ID']
os.environ['AWS_SECRET_ACCESS_KEY'] = config['AWS']['AWS_SECRET_ACCESS_KEY']

amazon_path = config['INPUT']['AMAZON_REVIEWS'] 
book_depository_path = config['INPUT']['BOOK_DEPOSITORY']
    
data_lake_path = config['OUTPUT']['DATA_LAKE_PATH']

## SparkSession

Creating the SparkSession:

In [3]:
spark = SparkSession \
    .builder \
    .appName("Books Data Lake") \
    .config("fs.s3a.awsAccessKeyId", os.environ['AWS_ACCESS_KEY_ID'])\
    .config("fs.s3a.awsSecretAccessKey", os.environ['AWS_SECRET_ACCESS_KEY'])\
    .config("spark.sql.csv.parser.columnPruning.enabled", False) \
    .config("spark.jars.packages", "org.apache.hadoop:hadoop-aws:2.7.1") \
    .getOrCreate() 

# Step 1: Scope the Project and Gather Data

## 1.1 Scope

[Book Depository](https://www.bookdepository.com/) is a leading international book retailer that has a wide book catalog and delivers worldwide. We are going to use a dataset of books metadata scraped from their site and made available on [Kaggle](https://www.kaggle.com/datasets). 

[Amazon](https://www.amazon.com/) has 20 years of online bookselling experience. Having always maintained high emphasis on customer satisfaction, the reliability of its Customer Reviews is one of the main strengths of the company. The [Amazon Customer Reviews Dataset](https://registry.opendata.aws/amazon-reviews/) is maintained by Amazon itself, and contains product reviews in multiple languages and product categories, including books. However, it focus exclusively on the reviews, and does not contain specific books metadata, like authors, edition and others. 

This project joins both datasets through the books International Standard Book Number (ISBN) producing a **Book Reviews Data Lake** hosted on S3, aimed to provide data for analytical purposes. We are going to use [Spark](https://spark.apache.org/) to build a data pipeline that consumes raw books metadata and Amazon product reviews from S3 buckets, processes the data into analytics tables using Spark in an EMR cluster, and writes those tables as parquet files into the S3 bucket that will host our data lake.

![Architecture Diagram](images/architecture.png)

## 1.2 Use Cases

Some intended use cases for this Data Lake:
* Sentiment analysis on book reviews;
* Understanding how people rate usefulness of a review of a book, and what factors influence helpfulness of a review;
* Feeding data for a book recommender system.

## 1.3 Data Sources

### *1.3.1. Book Depository Dataset*

This dataset is a large collection of books metadata, like title, description, authors, category and others. It was scraped from [Book Depository](https://www.bookdepository.com/), a leading international book retailer. 
 
 The following list describes each file in the dataset:
* `authors.csv` (**703,666 rows**): author names, matching the `authors` column from `dataset.csv`; a single book in the dataset may have more than one author.
* `categories.csv` (**2,767 rows**): category names, matching the `categories` column from `dataset.csv`; a single book in the dataset may have more than one category.
* `formats.csv` (**59 rows**): format names, matching the `format` column from `dataset.csv`.
* `places.csv` (**7,410 rows**): publication places (city and country), matching the `publication-place` column from `dataset.csv`.
* `dataset.csv` (**1,153,986 rows**): the actual dataset containing the catalog of books.

Source: https://www.kaggle.com/sp1thas/book-depository-dataset


### *1.3.2. Amazon Customer Reviews Dataset*

This dataset is managed by [Amazon](https://www.amazon.com/) and contains over 130+ million customer reviews written from 1995 until 2015. For this project, we are interested only in the reviews for products of the category "Books". 

This dataset is available in two file formats: TSV (tab separated value) and Parquet. Since Parquet is more efficient and is already partitioned by `product_category`, which makes it more convenient to use since we will be reading data specific to the "Books" category, we will be using only the Parquet files where `product_category = Books` fot this project. 

Number of rows in the category "Books" of this dataset: **20,726,160**

Source: https://registry.opendata.aws/amazon-reviews/

## 1.4. Reading the Data

In [4]:
authors_staging_df = spark.read.csv("{}/authors.csv".format(book_depository_path),
                                    quote='\"',
                                    escape='\"',
                                    sep=",",
                                    header=True, 
                                    inferSchema= True,
                                    multiLine=True)

In [10]:
categories_staging_df = spark.read.csv("{}/categories.csv".format(book_depository_path),
                                       quote='\"',
                                       escape='\"',
                                       sep=",",
                                       header=True, 
                                       inferSchema= True,
                                       multiLine=True)

In [16]:
formats_staging_df = spark.read.csv("{}/formats.csv".format(book_depository_path),
                                    quote='\"',
                                    escape='\"',
                                    sep=",",
                                    header=True, 
                                    inferSchema= True,
                                    multiLine=True)

In [22]:
places_staging_df = spark.read.csv("{}/places.csv".format(book_depository_path),
                                   quote='\"',
                                   escape='\"',
                                   sep=",",
                                   header=True, 
                                   inferSchema= True,
                                   multiLine=True)

In [28]:
books_staging_df = spark.read.csv("{}/dataset.csv".format(book_depository_path),
                                  quote='\"',
                                  escape='\"',
                                  sep=",",
                                  header=True, 
                                  inferSchema= True,
                                  multiLine=True)

In [34]:
amazon_staging_df = spark.read.parquet(amazon_path)

# Step 2: Explore and Assess the Data

In this step, we are going to explore the Spark DataFrames created in the previous section, in order to identify data quality issues and document steps necessary to clean the data.

### *2.1. Authors*

In [5]:
authors_staging_df.printSchema()

root
 |-- author_id: integer (nullable = true)
 |-- author_name: string (nullable = true)



In [6]:
authors_staging_df.limit(50).toPandas().head(50)

author_id                                        author_name
0        4872                                               None
1      374796                               !Mediengruppe Bitnik
2      295624                                           !ralupop
3      172310                                      # House Press
4       59149                               #1 Blank Comic Books
5      100888                                            #Value!
6      287590                                  #Worldlcass Media
7      552132                               #emergingproud Press
8      679802                              #shakeback Publishing
9       28924                                      #vettech Life
10     131732                                      & Bacon Allyn
11     668658                             & Berry Roskin & Berry
12     525351                & Beth Donelson Tom & Beth Donelson
13     457543                        & Bonchek Shepsle & Bonchek
14     490634                                    & Bradley Stone
15     668672                & Christensen Gerston & Christensen
16     668676  & Cook & Richardson-Gibbs Klein & Cook & Richa...
17     456132                             & Eagan Scarre & Eagan
18     668675                              & Feist Feist & Feist
19     700273                                        & Fun Books
20     588170                                    & Gamble Gamble
21     456983                         & Ginsberg Lowi & Ginsberg
22     668671                                 & Hess Harr & Hess
23     668673                     & Hoefler Fritschler & Hoefler
24     669342  & Holton III & Swanson Knowles & Holton III & ...
25     351786                                  & James Robertson
26     661268              & Kevin Lumsden David & Kevin Lumsden
27     188956                                     & Lange Carter
28     205360            & Leading Experts from Around the World
29     667954                        & Lepgold Lamborn & Lepgold
30      38539                            & Other Leading Experts
31     668680                            & Page Greenberg & Page
32     692964                        & Rrt Exam Secrets Test Crt
33     457545                             & Smith Baylis & Smith
34     216666                             & Southerdon Havenhand
35     586680                                & Stewart Smallwood
36     625318                      & Stigliani Spiro & Stigliani
37     668637  & Strasser & Rosenzweig Lichtenstein & Strasse...
38     456130                & Strickland Despelder & Strickland
39     456554                                 & Swim Herr & Swim
40     409709                                    & Urruty Zagier
41     328857                               & Wine Magazine Food
42     152255                                       & Zullo Nash
43     572855                                            &BEYOND
44     686870                                      &zoo Journals
45     696431                                         &zoo Press
46     423096                            'Abd Al-Qadir Al-Jilani
47     438588                           'Abd Al-Rahman Al-Jaziri
48     419073                      'Abd Allah B 'Umar Al-Baydawi
49     436192                          'Abd ar-Rahman Al-Akhdari

In [7]:
authors_staging_df.describe("author_id", "author_name").show()

+-------+-----------------+--------------------+
|summary|        author_id|         author_name|
+-------+-----------------+--------------------+
|  count|           703666|              703665|
|   mean|         351833.5|   8.151614933835E11|
| stddev|203131.0215973056|2.823802163249373E12|
|    min|                1|!Mediengruppe Bitnik|
|    max|           703666|              張洪聲|
+-------+-----------------+--------------------+



Checking the number of authors in the dataset:

In [8]:
authors_staging_df.count()

703666

Checking the number of authors in the dataset *removing duplicates*:

In [9]:
authors_staging_df.dropDuplicates().count()

703666

**Assessment:**

From the above assessment, we can see that the `author_id` column is clean, all values are numeric. 

Some `author_name` begin with non readable characters, but since it is not possible to identify which ones are valid authors an which are not, we are going to load them as they are, removing only the null values.

We can see also that there are no duplicated rows in the file.

### *2.2. Categories*

In [11]:
categories_staging_df.printSchema()

root
 |-- category_id: integer (nullable = true)
 |-- category_name: string (nullable = true)



In [12]:
categories_staging_df.limit(20).toPandas().head(20)

category_id                                category_name
0          1998                             .Net Programming
1           176  20th Century & Contemporary Classical Music
2          3291  20th Century & Contemporary Classical Music
3          2659      20th Century History: C 1900  To C 2000
4          2661          21st Century History: From C 2000 -
5          1992               2D Graphics: Games Programming
6          1943                      3D Graphics & Modelling
7          1993               3D Graphics: Games Programming
8          2472                                          ABC
9           768                          Abnormal Psychology
10         2023                                       Access
11         1379                Accident & Emergency Medicine
12         1422                 Accident & Emergency Nursing
13          957                                   Accounting
14         1932                          Accounting Software
15          963          Accounting: Study & Revision Guides
16         1856                 Acoustic & Sound Engineering
17          391        Acquisitions & Collection Development
18          127                            Acting Techniques
19         3091                      Active Outdoor Pursuits

In [13]:
categories_staging_df.describe("category_id", "category_name").show()

+-------+------------------+----------------+
|summary|       category_id|   category_name|
+-------+------------------+----------------+
|  count|              2767|            2767|
|   mean|1709.2696060715577|            null|
| stddev|1004.8571823557746|            null|
|    min|                 2|.Net Programming|
|    max|              3390|  Zoroastrianism|
+-------+------------------+----------------+



Checking the number of book categories in the dataset:

In [14]:
categories_staging_df.count()

2767

Checking the number of book categories in the dataset *removing duplicates*:

In [15]:
categories_staging_df.dropDuplicates().count()

2767

**Assessment:**

The categories data looks clean. We are going to load them as they are. We can see also that there are no duplicated rows in the file.

### *2.3. Formats*

In [17]:
formats_staging_df.printSchema()

root
 |-- format_id: integer (nullable = true)
 |-- format_name: string (nullable = true)



In [18]:
formats_staging_df.limit(20).toPandas().head(20)

format_id   format_name
0          17       Address
1          12         Audio
2          50          Bath
3          53       Betamax
4          51           Big
5          26         Board
6          13          Book
7          41         Boxed
8           7            CD
9           4      Calendar
10         10         Cards
11         36   Copymasters
12         33   Counterpack
13         44           DAT
14         28           DVD
15         11         Diary
16         37       Digital
17         45      Diskette
18          9  Downloadable
19         48       Dumpbin

In [19]:
formats_staging_df.describe("format_id", "format_name").show()

+-------+------------------+-----------+
|summary|         format_id|format_name|
+-------+------------------+-----------+
|  count|                59|         59|
|   mean|              30.0|       null|
| stddev|17.175564037317667|       null|
|    min|                 1|    Address|
|    max|                59|  Wallchart|
+-------+------------------+-----------+



Checking the number of book formats in the dataset:

In [20]:
formats_staging_df.count()

59

Checking the number of book formats in the dataset *removing duplicates*:

In [21]:
formats_staging_df.dropDuplicates().count()

59

**Assessment:**

The formats data also looks clean, and will be loaded without changes. We can see also that there are no duplicated rows in the file.

### *2.4. Publication Places*

In [23]:
places_staging_df.printSchema()

root
 |-- place_id: integer (nullable = true)
 |-- place_name: string (nullable = true)



In [24]:
places_staging_df.limit(50).toPandas().head(50)

place_id                                    place_name
0       5996  (56 Bourne Way, Hayes, Kent), United Kingdom
1        619                             ), United Kingdom
2       4307                              0, United States
3       3488                            189, United States
4       7326                                           192
5       4601                            204, United States
6       4602                            205, United States
7       5234                                           206
8       4600                            206, United States
9       5233                            216, United States
10       849                            237, United States
11      6269                            241, United States
12      4555                                           258
13      2213                           75004 Paris, France
14      4936                           75191 Paris, France
15      3492                               A Coruna, Spain
16      5262                       A Coruna, United States
17      6579                                    AB, Canada
18      4055                             AL, United States
19      6767                    APSE HEATH, United Kingdom
20       193                             AR, United States
21      4743                      ARDLEIGH, United Kingdom
22      2083                               AX, Netherlands
23      1645                             AZ, United States
24       914                               Aachen, Germany
25      5480                            Aarau, Switzerland
26       109                               Aarhus, Denmark
27      3040                      Abberton, United Kingdom
28      2244                      Abbeville, United States
29      6635                Abbots Langley, United Kingdom
30      3567                     Abbotsford, United States
31      4677                    Abbotsford, VIC, Australia
32      2290                   Aberdeen, NJ, United States
33       665                      Aberdeen, United Kingdom
34      5024                 Aberdeenshire, United Kingdom
35      5332                     Aberfeldy, United Kingdom
36      3473                      Abergale, United Kingdom
37      4777                   Abergavenny, United Kingdom
38      3637                      Abergele, United Kingdom
39      6475                  Aberglasslyn, NSW, Australia
40      1929                     Aberporth, United Kingdom
41      1971                   Abertillery, United Kingdom
42      1860                   Aberystwyth, United Kingdom
43      2091                                   Abilene, TX
44      3504                    Abilene, TX, United States
45      2679                        Abilene, United States
46       186                      Abingdon, United Kingdom
47      2300                      Abington, United Kingdom
48      2729                       Abington, United States
49      6782                        Aboyne, United Kingdom

In [25]:
places_staging_df.describe("place_id", "place_name").show()

+-------+------------------+--------------------+
|summary|          place_id|          place_name|
+-------+------------------+--------------------+
|  count|              7410|                7410|
|   mean|            3705.5|  218.66666666666666|
| stddev|2139.2270800455003|    34.7754702819866|
|    min|                 1|(56 Bourne Way, H...|
|    max|              7410|Москва, United St...|
+-------+------------------+--------------------+



Checking the number of publication places in the dataset:

In [26]:
places_staging_df.count()

7410

Checking the number of publication places in the dataset *removing duplicates*:

In [27]:
places_staging_df.dropDuplicates().count()

7410

**Assessment:**

The `place_name` column has some issues, like numeric characters at the begining of the string. To clean it, we will make some replacements during the ETL.

We can see that there are no duplicated rows in the file.

### *2.5. Books*

In [29]:
books_staging_df.printSchema()

root
 |-- authors: string (nullable = true)
 |-- bestsellers-rank: string (nullable = true)
 |-- categories: string (nullable = true)
 |-- description: string (nullable = true)
 |-- dimension-x: double (nullable = true)
 |-- dimension-y: double (nullable = true)
 |-- dimension-z: double (nullable = true)
 |-- edition: string (nullable = true)
 |-- edition-statement: string (nullable = true)
 |-- for-ages: string (nullable = true)
 |-- format: string (nullable = true)
 |-- id: long (nullable = true)
 |-- illustrations-note: string (nullable = true)
 |-- imprint: string (nullable = true)
 |-- index-date: timestamp (nullable = true)
 |-- isbn10: string (nullable = true)
 |-- isbn13: long (nullable = true)
 |-- lang: string (nullable = true)
 |-- publication-date: timestamp (nullable = true)
 |-- publication-place: integer (nullable = true)
 |-- rating-avg: double (nullable = true)
 |-- rating-count: integer (nullable = true)
 |-- title: string (nullable = true)
 |-- url: string (nullable 

In [30]:
books_staging_df.limit(50).toPandas().head(50)

authors bestsellers-rank  \
0     [1, 2]             2896   
1        [3]            51517   
2        [4]           162985   
3        [5]            41520   
4        [6]             3885   
5        [7]            75224   
6     [8, 9]           368836   
7       [10]          108,369   
8   [11, 12]            49553   
9   [13, 14]          128,070   
10  [15, 16]          148,229   
11      [17]           454813   
12      [18]           171512   
13      [19]           493920   
14      [20]           69,927   
15      [21]           536410   
16      [22]           204467   
17      [23]           583236   
18      [24]          223,872   
19      [25]          1091127   
20      [26]           201409   
21      [27]            33498   
22      [28]           299341   
23      [29]          1257939   
24      [20]          258,448   
25      [30]          1465502   
26      [31]            71921   
27  [32, 33]          1343111   
28  [34, 35]            79024   
29      [36]          1964763   
30      [37]           322144   
31  [38, 39]        2,122,899   
32      [40]           364771   
33      [41]          1879742   
34  [42, 43]           91,936   
35      [44]            86798   
36      [45]          1893170   
37      [46]            82795   
38      [47]            28176   
39  [48, 49]          2177559   
40      [50]          306,379   
41      [51]           80,718   
42      [52]          2157077   
43      [53]           408788   
44      [54]             None   
45      [20]          421,308   
46      [55]           214900   
47      [56]          2698437   
48  [57, 58]           391619   
49      [59]           246529   

                                           categories  \
0                               [235, 3386, 736, 763]   
1                                         [360, 2632]   
2                            [1703, 2771, 2818, 3097]   
3                             [819, 3364, 1853, 2977]   
4                                  [1694, 1703, 2818]   
5                                  [1843, 2967, 2969]   
6                      [1520, 2771, 3019, 3074, 3078]   
7                                  [1853, 2977, 3023]   
8                             [649, 2771, 3019, 3074]   
9                            [1725, 1844, 2969, 2973]   
10                                 [1711, 1843, 2969]   
11         [1379, 1701, 1703, 1738, 2818, 2884, 3097]   
12                                 [1844, 2969, 3058]   
13                           [1691, 1694, 1703, 2818]   
14                     [1725, 1844, 2969, 2971, 2973]   
15                                 [2818, 2968, 3063]   
16                                 [1849, 2975, 3069]   
17         [2813, 2818, 2820, 2987, 3091, 3092, 3100]   
18                           [1843, 2802, 2969, 2973]   
19                                       [2771, 3097]   
20                                 [1719, 1843, 2969]   
21                                       [1703, 2818]   
22                                 [1850, 3071, 3385]   
23                           [2771, 3019, 3074, 3076]   
24                      [376, 1725, 1844, 2971, 2973]   
25                              [376, 378, 888, 2771]   
26                                       [1848, 3069]   
27                                 [2770, 2771, 2772]   
28                     [1755, 1759, 1762, 1849, 2975]   
29           [840, 862, 1289, 1298, 1303, 2771, 2772]   
30                                             [1844]   
31                             [376, 378, 2771, 3097]   
32                           [1850, 2975, 3069, 3071]   
33                           [2771, 3019, 3020, 3074]   
34                     [1725, 1755, 1848, 1867, 2975]   
35   [2659, 2663, 2858, 2859, 2868, 2878, 2978, 2915]   
36  [214, 231, 233, 2777, 2782, 2804, 2813, 2818, ...   
37                                 [1477, 1850, 3071]   
38                           [2797, 2800, 2801, 2915]   
39                                       

In [31]:
books_staging_df.describe().toPandas().head()

summary  authors    bestsellers-rank             categories  \
0   count  1153986              461288                1153986   
1    mean     None  1073662.3132837743                   None   
2  stddev     None    842385.808938867                   None   
3     min   [1, 2]           1,000,130  [10, 103, 2902, 2954]   
4     max       []              999997                     []   

                                         description        dimension-x  \
0                                            1042082            1101652   
1                                 1014.1333333333333  163.5892014538665   
2                                  975.7049368582306  36.65227604035362   
3  ! THIS BOOK IS INCLUDED IN ITALIAN CUISINE COO...               0.25   
4  ����★☆ What Is Stopping You From Making Money ...             3962.0   

          dimension-y        dimension-z   edition         edition-statement  \
0             1051356            1101652    123180                    306720   
1  232.09219199780216   25.5515790013423      None        1648.2755432269996   
2   39.53291989305113  50.42905419090038      None        2366.4170096067755   
3                 1.0                0.1  Abridged          #5 of Series ed.   
4              2000.0             2000.0  Variorum  überarbeitete Neuauflage   

  for-ages                                            format  \
0    55344                                           1147715   
1     None                                2.0372931410631616   
2     None                                 3.315326294695639   
3     0-10                                                 1   
4     9-12  Wallchart\n                                    |   

                     id illustrations-note       imprint  \
0               1153986             394403        228849   
1  9.781778054683115E12  208.5388974983396     9543261.0   
2  1.7700559289798207E9  594.1026705181978           NaN   
3         9780000005199                  "         12...   
4         9798389450189       über 80 Abb.  united p. c.   

                  isbn10                isbn13     lang   publication-place  \
0                1153986               1138650  1078206              737199   
1  1.4448580669936057E11  9.781784246509844E12     None  337.11218680437713   
2  1.1726851144641418E12  1.7738776960178776E9     None   796.9126703460335   
3             000017758X         9780000005199       ab                   1   
4             9999991319         9798389450189      zxx                7410   

           rating-avg       rating-count  \
0              486376             486376   
1  3.9514950162016254  5177.365287349705   
2   0.622674740999877  76965.97854149972   
3                 1.0                  1   
4                 5.0            6473716   

                                           title  \
0                                        1153986   
1                                       Infinity   
2                                            NaN   
3                          ! : New Soups dieting   
4  �tudes Et Lectures Sur l'Astronomie. Tome 4   

                                                 url              weight  
0                                            1153986             1059685  
1                                               None  475.52386219496674  
2                                               None    716.884990113177  
3  /-1044-1072-1088-1100-1103-1053-1077-1089-1090...                15.0  
4            /zzSongs-Verse-Dahl-Roald/9780141356136             82522.0

Checking the number of books in the dataset:

In [32]:
books_staging_df.count()

1153986

Checking the number of books in the dataset *removing duplicates*:

In [33]:
books_staging_df.dropDuplicates().count()

1153986

**Assessment:**

The Books table has a many-to-many relationship with the entities `authors` and `categories`. In the dataset, these relationships is represented through lists of IDs (e.g. "[235, 3386, 736, 763]"). Because of that, we will create two aditional tables to serve as many-to-many bridges between Books-Author and Books-Category.

We can see also that there are no duplicated rows in the Books file.

### *2.6. Amazon Reviews*

In [35]:
amazon_staging_df.printSchema()

root
 |-- marketplace: string (nullable = true)
 |-- customer_id: string (nullable = true)
 |-- review_id: string (nullable = true)
 |-- product_id: string (nullable = true)
 |-- product_parent: string (nullable = true)
 |-- product_title: string (nullable = true)
 |-- star_rating: integer (nullable = true)
 |-- helpful_votes: integer (nullable = true)
 |-- total_votes: integer (nullable = true)
 |-- vine: string (nullable = true)
 |-- verified_purchase: string (nullable = true)
 |-- review_headline: string (nullable = true)
 |-- review_body: string (nullable = true)
 |-- review_date: date (nullable = true)
 |-- year: integer (nullable = true)



In [36]:
amazon_staging_df.limit(20).toPandas().head(20)

marketplace customer_id       review_id  product_id product_parent  \
0           US    15444933  R1WWG70WK9VUCH  1848192576      835940987   
1           US    20595117  R1EQ3POS0RIOD5  145162445X      574044348   
2           US    52925878  R10SRJA4VVGUBD  055341805X      846590203   
3           US    40062567   RD3268X41GM7U  0425263908      119148606   
4           US    47221489  R3KGQL5X5BSJE1  1416556141      987400385   
5           US    20709973  R3IY9A2Z7N2R49  1455523003      181218559   
6           US    18052694  R12PGYPH9B9UVJ  0990388697      840404589   
7           US    48279565  R2CA9LLJBEORNZ  0345803485      600633062   
8           US    14297627  R2C8BZMS1JRZ04  1469245167      466321651   
9           US    37340439  R2F8UI9NNP6S33  0345803485      600633062   
10          US    35122979  R18L06ZV927ZS0  0743477111       43217624   
11          US    11213632  R32NFRKUPJ3E4J  1609611543      657488375   
12          US    15816241  R1ZQ0M1HIMV8T9  0563551933      412573914   
13          US    36748944   RKEHECW1IR9DT  0061777269       84797790   
14          US    38261190  R3HKUZRLPR36SW  0316769487      862964341   
15          US    44704475  R36X05VUTXCEJQ  0307730697       62706984   
16          US    38351306  R257VQLC12S8IJ  0448439158       27871958   
17          US    49320969  R1DA0O932TXFV7  0375842209       47504452   
18          US    21239776   RAF45VHW4W7HU  1402773536      615203301   
19          US    39150739  R1YBB6HXVCZ8AR  0307592731      576844294   

                                        product_title  star_rating  \
0   Standing Qigong for Health and Martial Arts - ...            5   
1   A Universe from Nothing: Why There Is Somethin...            4   
2                             Hyacinth Girls: A Novel            4   
3                                        Bared to You            5   
4                                     Healer: A Novel            5   
5   The Missionary Position: Mother Teresa in Theo...            4   
6               I'm Tired of Being Ordinary, Are You?            4   
7   Fifty Shades of Grey: Book One of the Fifty Sh...            1   
8                               The Thrill of Victory            4   
9   Fifty Shades of Grey: Book One of the Fifty Sh...            5   
10      Romeo and Juliet (Folger Shakespeare Library)            4   
11  Wheat Belly: Lose the Wheat, Lose the Weight, ...            5   
12  Dangerous Desserts: 200 Tried-and-tested Recip...            5   
13  Consciousness Beyond Life: The Science of the ...            5   
14                             The Catcher in the Rye            1   
15  Fearless: The Undaunted Courage and Ultimate S...            4   
16                               Best-Ever Big Sister            5   
17                                     The Book Thief            5   
18                                 Large Print Sudoku            5   
19  Wild: From Lost to Found on the Pacific Crest ...            5   

    helpful_votes  total_votes vine verified_purchase  \
0               9           10    N                 Y   
1               4            7    N                 N   
2               0            0    Y                 N   
3               1            1    N                 N   
4               0            0    N                 Y   
5               5            7    N                 Y   
6               1            1    N                 Y   
7               7            7    N                 N   
8               0            0    N                 Y   
9               0            9    N                 N   
10              0            0    N                 Y   
11              1            1    N                 Y   
12              0            0    N                 Y   
13              0            0    N                 N   
14              1            6    N                 Y   
15              0            0    N                 Y   
16              0            

Checking the number of reviews in the dataset:

In [37]:
amazon_staging_df.count()

20726160

Checking the number of reviews in the dataset *removing duplicates*:

In [38]:
amazon_staging_df.dropDuplicates().count()

19887359

Checking for reviews in invalid (future) dates:

In [39]:
invalid_reviews = amazon_staging_df.filter(f.col("review_date") > f.unix_timestamp(timestamp=None,format='yyyy-MM-dd HH:mm:ss').cast('timestamp'))

invalid_reviews.count()

0

**Assessment:**

* This dataset **has duplicated rows** that will have to be eliminated in the ETL. 

* There are no rows with `review_date` greater than the current timestamp.

# Step 3: Define the Data Model

The Data Lake built by this project is meant to be used mainly for analytic queries (see section *1.2 Use Cases*). Therefore, the schema chosen for the data model was the **snowflake schema**. 

The fact table was defined to be the table containing the Amazon reviews. The dimension tables defined for the schema are the **dates** when the reviews were written and the **books** metadata, which in turn has its own related tables: **authors**, **categories**, **formats** and **publication places**.

Also, since the books table has a many-to-many relationship with the entities authors and categories, two aditional tables were defined to serve as bridges between them: **books-authors** and **books-categories**.

After exploring the datasets, the resulting Conceptual Data Model created was this:

![Data Model](images/data_model.png)

# Step 4: Run Pipelines to Model the Data 

## 4.1 Create the data model

In this step, the data will be cleaned and loaded into the tables, creating the data model defined in the previous section.

### *4.1.1. Authors Table*

In [40]:
# Registering the DataFrame as a SQL temporary view
authors_staging_df.createOrReplaceTempView("authors_staging_table")

# Loading the table
authors_table = spark.sql(
        """
        SELECT CAST(author_id AS int) AS author_id,
               author_name
        FROM authors_staging_table
        WHERE author_name IS NOT NULL
        """)

# Checking results
authors_table.sort("author_id").limit(10).toPandas().head(10)

author_id        author_name
0          1       John Douglas
1          2      Mark Olshaker
2          3      Cecelia Mecca
3          4   Ana Maria Spagna
4          5  Mark Vanhoenacker
5          6           Jen Gale
6          7      Giles Chapman
7          8          Sam Nigro
8          9      J. D. Swanson
9         10     Brian Cosgrove

### *4.1.2. Categories Table*

In [41]:
# Registering the DataFrame as a SQL temporary view
categories_staging_df.createOrReplaceTempView("categories_staging_table")

# Loading the table
categories_table = categories_staging_df

# Checking results
categories_table.sort("category_id").limit(10).toPandas().head(10)

category_id      category_name
0            2  Art & Photography
1            3          Art Books
2            4         Art Theory
3            5     Art Resoration
4            6            Forgery
5            7        Art Finance
6            8        Art History
7            9             Styles
8           10     Indigenous Art
9           11          Naive Art

### *4.1.3. Formats Table*

In [42]:
# Register the DataFrame as a SQL temporary view
formats_staging_df.createOrReplaceTempView("formats_staging_table")

# Loading the table
formats_table = formats_staging_df

# Checking results
formats_table.sort("format_id").limit(10).toPandas().head(10)

format_id   format_name
0          1     Paperback
1          2      Hardback
2          3        Spiral
3          4      Calendar
4          5      Pamphlet
5          6          Game
6          7            CD
7          8      Notebook
8          9  Downloadable
9         10         Cards

### *4.1.4. Publication Places Table*

In [43]:
# Registering the DataFrame as a SQL temporary view
places_staging_df.createOrReplaceTempView("places_staging_table")

# Loading the table
places_table = spark.sql(
        """
        SELECT place_id,
               CASE WHEN place_name rlike '^[0-9]*$' THEN null
                    WHEN place_name rlike '^[0-9]+ Paris, France$' THEN 'Paris, France'
                    WHEN place_name rlike '^[0-9]+[,] United States$' THEN 'United States'
                    WHEN place_name rlike '[\)], United Kingdom' THEN 'United Kingdom'                                       
                    ELSE place_name 
                    END AS place_name
        FROM places_staging_table
        WHERE place_name is not null AND not place_name rlike '^[0-9]*$'
        """)

# Checking results
places_table.sort("place_id").limit(10).toPandas().head(10)

place_id                    place_name
0         1        London, United Kingdom
1         2    North Adams, United States
2         3      Wisconsin, United States
3         4  Rolindale, MA, United States
4         5      Ramsbury, United Kingdom
5         6      Somerset, United Kingdom
6         7   New York, NY, United States
7         8   Rockport, ME, United States
8         9       New York, United States
9        10                 United States

### *4.1.5. Books Table*

In [44]:
# Registering the DataFrame as a SQL temporary view
books_staging_df.createOrReplaceTempView("books_staging_table")

# Loading the table
books_table = spark.sql(
        """
        SELECT id as book_id,
               isbn10,
               isbn13,
               title,
               authors as authors_list,
               categories as categories_list,               
               edition,
               lang,
               format as format_id,
               `publication-place` as publication_place_id,
               `publication-date` as publication_date,
               description
        FROM books_staging_table
        """)

# Checking results
books_table.limit(50).toPandas().head(50)

book_id         isbn10        isbn13  \
0   9781787465145     1787465144  9.781787e+12   
1   9781946510402     1946510408  9.781947e+12   
2   9781612124568     1612124569  9.781612e+12   
3   9781786487155     1786487152  9.781786e+12   
4   9781472969125     147296912X  9.781473e+12   
5   9780760363997      760363994  9.780760e+12   
6   9781594394591     1594394598  9.781594e+12   
7   9781847975096     1847975097  9.781848e+12   
8   9781594396816     1594396817  9.781594e+12   
9   9781563924699  9781563924699           NaN   
10  9780760336953      760336954  9.780760e+12   
11  9781632206343     163220634X  9.781632e+12   
12  9781781316283     1781316287  9.781781e+12   
13  9780740779336      740779338  9.780741e+12   
14  9780856961250      856961256  9.780857e+12   
15  9780143111771      143111779  9.780143e+12   
16  9780071464055       71464050  9.780071e+12   
17  9781912023646     1912023644  9.781912e+12   
18  9781557885586     1557885583  9.781558e+12   
19  9781440247286     1440247285  9.781440e+12   
20  9780571248117     057124811X  9.780571e+12   
21  9781977954749     197795474X  9.781978e+12   
22  9781898660354     1898660352  9.781899e+12   
23  9781594396557     1594396558  9.781594e+12   
24  9780856965210  9780856965210           NaN   
25  9780060735203       60735201  9.780061e+12   
26  9780957311916      957311915  9.780957e+12   
27  9780425211946      425211940  9.780425e+12   
28  9780521559263     052155926X  9.780522e+12   
29  9780062407535       62407538  9.780062e+12   
30  9780435048150      435048155  9.780435e+12   
31  9781493023769  9781493023769           NaN   
32  9781472918130     1472918134  9.781473e+12   
33  9781594396007     1594396000  9.781594e+12   
34  9780080972398     008097239X  9.780081e+12   
35  9781910931387     1910931381  9.781911e+12   
36  9781784521554     1784521558  9.781785e+12   
37  9780470518793      470518790  9.780471e+12   
38  9780307888259      307888258  9.780308e+12   
39  9781530784530     1530784530  9.781531e+12   
40  9788879117333     8879117335  9.788879e+12   
41  9780717149773      717149773  9.780717e+12   
42  9781438004082     1438004087  9.781438e+12   
43  9781557885456     1557885451  9.781558e+12   
44  9781608761784     1608761789  9.781609e+12   
45  9780856963391      856963399  9.780857e+12   
46  9781910374337     1910374334  9.781910e+12   
47  9781448878666     1448878667  9.781449e+12   
48  9780803235342      803235348  9.780803e+12   
49  9781910602829     1910602825  9.781911e+12   

                                                title authors_list  \
0                               Journey Into Darkness       [1, 2]   
1           The Mercenary : Order of the Broken Blade          [3]   
2   100 Skills You'll Need for the End of the Worl...          [4]   
3                                 How to Land a Plane          [5]   
4   The Sustainable(ish) Living Guide : Everything...          [6]   
5                                     Mini : 60 Years          [7]   
6                   Karate Science : Dynamic Movement       [8, 9]   
7                         Microlight Pilot's Handbook         [10]   
8               Krav Maga Combatives : Maximum Effect     [11, 12]   
9   Dodge Caravan, Plymouth Voyager & Chrysler Tow...     [13, 14]   
10                      How to Design Cars Like a Pro     [15, 16]   
11  52 Unique Techniques for Stocking Food for Pre...         [17]   
12                                Land Rover Defender         [18]   
13  Sneaky Green Uses for Everyday Things : How to...         [19]   
14                          Norton Commando (68 - 77)         [20]   
15                     How Cycling Can Save the World         [21]   
16                       The Boatbuilder's Apprentice         [22]   
17  Thinking on My Feet : The small joy of putting...         [23]   
18               Ford Windsor Small-block Performance         [24]   
19  365 Essential Survival Skills : Knowledge that...      

### *4.1.6. Book-Categories Relationship Table*

The Books table has a many-to-many relationship with the Categories Table. In the original dataset, these relationships is represented through lists of IDs. Because of that, we will we will parse these lists in order to create an aditional table to serve as a many-to-many bridge between Books and Categories.

In [45]:
# Defining the parsing function
def convert_string_to_list(s):
    return s.strip('][').split(', ')

convert_string_to_list_udf = f.udf(lambda s: convert_string_to_list(s), t.ArrayType(t.StringType()))

# Loading the table
books_categories_table = books_staging_df.select(
    f.col("id").alias("book_id"), 
    f.explode(convert_string_to_list_udf("categories")).alias("category_id")
)

# Checking results
books_categories_table.limit(20).toPandas().head(20)

book_id category_id
0   9781787465145         235
1   9781787465145        3386
2   9781787465145         736
3   9781787465145         763
4   9781946510402         360
5   9781946510402        2632
6   9781612124568        1703
7   9781612124568        2771
8   9781612124568        2818
9   9781612124568        3097
10  9781786487155         819
11  9781786487155        3364
12  9781786487155        1853
13  9781786487155        2977
14  9781472969125        1694
15  9781472969125        1703
16  9781472969125        2818
17  9780760363997        1843
18  9780760363997        2967
19  9780760363997        2969

### *4.1.7. Book-Authors Relationship Table*

The Books table has a many-to-many relationship with the Authors Table. In the original dataset, these relationships is represented through lists of IDs. Because of that, we will parse these lists in order to create an aditional table to serve as a many-to-many bridge between Books and Authors.

In [46]:
# Loading the table
books_authors_table = books_staging_df.select(
    f.col("id").alias("book_id"), 
    f.explode(convert_string_to_list_udf("authors")).alias("author_id")
)

# Checking results
books_authors_table.limit(20).toPandas().head(20)

book_id author_id
0   9781787465145         1
1   9781787465145         2
2   9781946510402         3
3   9781612124568         4
4   9781786487155         5
5   9781472969125         6
6   9780760363997         7
7   9781594394591         8
8   9781594394591         9
9   9781847975096        10
10  9781594396816        11
11  9781594396816        12
12  9781563924699        13
13  9781563924699        14
14  9780760336953        15
15  9780760336953        16
16  9781632206343        17
17  9781781316283        18
18  9780740779336        19
19  9780856961250        20

### *4.1.8. Amazon Reviews Table*

For books in Amazon, the `product_id` corresponds to the books's ISBN-10. Therefore, we are joining the Amazon Reviews data and the Books Depository metadata through the ISBN-10.

We are going to use `INNER JOIN` to match books and reviews. Therefore, **reviews that are NOT associated with any row in the Books Table will not be loaded into the Data Lake.**

In [47]:
# Registering the Amazon Reviews Staging DataFrame as a SQL temporary view
amazon_staging_df.createOrReplaceTempView("amazon_reviews_staging_table")

# Registering the Books Table as a SQL temporary view
books_table.createOrReplaceTempView("books_table")

# Removing the duplicated rows
amazon_staging_df = amazon_staging_df.dropDuplicates()

# Loading Amazon Reviews Table, by joining the Amazon Reviews Staging DataFrame and the Books Table.
# For books, the product_id is the books's ISBN
amazon_reviews_table = spark.sql(
    """
    SELECT a.review_id, 
           b.book_id, 
           a.star_rating, 
           a.helpful_votes, 
           a.total_votes, 
           a.verified_purchase, 
           a.customer_id,
           a.review_date, 
           year(a.review_date) as review_year, 
           a.review_headline, 
           a.review_body
    FROM amazon_reviews_staging_table a
         INNER JOIN books_table b ON a.product_id = b.isbn10
    """)

# Checking results
amazon_reviews_table.limit(20).toPandas().head(20)

review_id        book_id  star_rating  helpful_votes  total_votes  \
0   R1BFKJAX7PLZY2  9780060878078            4              6            9   
1   R24LW7WLO6UF6W  9780060878078            5             22           24   
2    RHDONITYOSLD7  9780060878078            1             11           18   
3   R2POBJH7MJ3KHL  9780060878078            3              7           10   
4    RKAKST5M35YVG  9780060878078            5              5            5   
5   R2W8J5L0K2BYXJ  9780060878078            1             13           27   
6   R202WY3U0RCYAN  9780060878078            4              0            0   
7   R1FG7OWVG4PJ9A  9780060878078            1              0            0   
8   R2R2A3TPJCBSEM  9780060878078            5              2            2   
9    R8PICYUMJUY10  9780060878078            5              3           13   
10    RE7BENULOQ8I  9780060878078            2              2            3   
11  R25MAKWV6PNZ50  9780060878078            4             11           11   
12  R3EBAOCYSG0ETQ  9780060878078            4              0            0   
13  R3SBOS4DBI8VMW  9780071406000            5             10           11   
14  R2IEBRRRHLHPUQ  9780071406000            5              0            2   
15  R1I9DY0SYMQ4TA  9780071406000            5             12           12   
16   RQCG4EW1SY94D  9780071406000            4              1            1   
17  R19NQJZBWS5W9K  9780091864958            3              5            5   
18  R1MBRWY86DD8GC  9780126930184            5              2            2   
19  R38R681R87HU4C  9780126930184            4              0            0   

   verified_purchase customer_id review_date  review_year  \
0                  N    14810040  2008-05-29         2008   
1                  N    40072440  2007-11-23         2007   
2                  N    17492715  2008-05-20         2008   
3                  N    27331930  2009-04-08         2009   
4                  N    34247947  2013-02-28         2013   
5                  Y    18058313  2007-11-15         2007   
6                  Y     2762668  2015-03-01         2015   
7                  N    52456601  2013-12-28         2013   
8                  Y    15813941  2012-12-02         2012   
9                  N    11229368  2006-02-20         2006   
10                 Y    40347942  2015-06-19         2015   
11                 N    52878657  2008-12-29         2008   
12                 Y    48435723  2013-07-21         2013   
13                 Y    52236461  2003-05-19         2003   
14                 Y    37916136  2012-01-14         2012   
15                 N    25453183  2006-02-27         2006   
16                 N    52044054  2010-08-21         2010   
17                 N    49775107  2005-09-04         2005   
18                 Y    19509353  2012-11-04         2012   
19                 Y     2419119  2014-08-26         2014   

                                      review_headline  \
0                      Informative but sadly negative   
1                           Interesting & fun to read   
2                   Popes through the eyes of dissent   
3   Solid Sketches of the Popes But Know What You ...   
4       A great overview of a fascinating institution   
5             Expected History and Got A Movie Review   
6                                          Four Stars   
7                   "The number of fools is infinite"   
8     An incredible history of an ancient institution   
9                                         Pretty Good   
10                            So so and nothing more.   
11                             A fascinating resource   
12         I was intrigued by the lives of the popes.   
13  Good overview of the possibilities for those w...   
14                                   EXCELLENT SELLER   
15                             A great place to begin   
16                   Great insight for writer careers   
17    Frustrating but it may suit your learning style   
18   

### *4.1.9. Date Table*

In [48]:
# Selecting the review dates from the Amazon Reviews Table, dropping the duplicated values
date_df = amazon_reviews_table.select("review_date").dropDuplicates()

# Loading the table
date_table = date_df.select(
    "review_date",
    f.year("review_date").alias('year'), 
    f.month("review_date").alias('month'), 
    f.dayofmonth("review_date").alias('day'),
    f.weekofyear("review_date").alias('week'),
    f.dayofweek("review_date").alias('weekday')
)

# Checking results
date_table.limit(10).toPandas().head(10)

review_date  year  month  day  week  weekday
0  2007-11-23  2007     11   23    47        6
1  2007-11-15  2007     11   15    46        5
2  2013-03-26  2013      3   26    13        3
3  2015-03-09  2015      3    9    11        2
4  2015-05-19  2015      5   19    21        3
5  2012-04-17  2012      4   17    16        3
6  2002-12-25  2002     12   25    52        4
7  2010-08-11  2010      8   11    32        4
8  2014-11-12  2014     11   12    46        4
9  2014-09-26  2014      9   26    39        6

## 4.2. Writing the tables into the Data Lake as parquet files

In [67]:
authors_table \
    .write \
    .mode("overwrite") \
    .parquet(os.path.join(data_lake_path, "authors.parquet"))  

In [68]:
categories_table \
    .write \
    .mode("overwrite") \
    .parquet(os.path.join(data_lake_path, "categories.parquet"))  

In [69]:
formats_table \
    .write \
    .mode("overwrite") \
    .parquet(os.path.join(data_lake_path, "formats.parquet"))  

In [70]:
places_table \
    .write \
    .mode("overwrite") \
    .parquet(os.path.join(data_lake_path, "places.parquet"))  

In [71]:
books_table \
    .write \
    .mode("overwrite") \
    .parquet(os.path.join(data_lake_path, "books.parquet"))  

In [72]:
books_categories_table \
    .write \
    .mode("overwrite") \
    .parquet(os.path.join(data_lake_path, "books_categories.parquet"))  

In [75]:
books_authors_table \
    .write \
    .mode("overwrite") \
    .parquet(os.path.join(data_lake_path, "books_authors.parquet"))  

In [77]:
amazon_reviews_table \
    .repartition("review_year") \
    .write \
    .partitionBy("review_year") \
    .mode("overwrite") \
    .parquet(os.path.join(data_lake_path, "amazon_reviews.parquet"))  

In [80]:
date_table \
    .repartition("year") \
    .write \
    .partitionBy("year") \
    .mode("overwrite") \
    .parquet(os.path.join(data_lake_path, "dates.parquet"))  

## 4.3 Data Quality Checks

### *4.3.1. Checking that none of the tables is empty*

In [49]:
authors_table.count()

703665

In [50]:
categories_table.count()

2767

In [51]:
formats_table.count()

59

In [52]:
places_table.count()

7407

In [53]:
books_table.count()

1153986

In [54]:
books_categories_table.count()

4074354

In [55]:
books_authors_table.count()

1600548

In [56]:
date_table.count()

6710

In [57]:
amazon_reviews_table.count()

1077713

### *4.3.2. Checking for unwanted NULL columns*

In [58]:
authors_table.filter("author_id is null OR author_name is null").count()

0

In [59]:
categories_table.filter("category_id is null or category_name is null").count()

0

In [60]:
formats_table.filter("format_id is null or format_name is null").count()

0

In [61]:
places_table.filter("place_id is null or place_name is null").count()

0

In [62]:
books_table.filter("book_id is null or isbn10 is null or title is null").count()

0

In [63]:
books_categories_table.filter("book_id is null or category_id is null").count()

0

In [64]:
books_authors_table.filter("book_id is null or author_id is null").count()

0

In [65]:
amazon_reviews_table.filter("review_id is null or book_id is null or review_date is null").count()

0

In [66]:
date_table.filter("review_date is null").count()

0

Closing Spark Session:

In [81]:
spark.stop()

## 4.4 Data dictionary 

In [4]:
%%html
<style>
table {float:left}
</style>

Table: **AUTHORS** \
Description: book authors
 
| Column | Type | Description |
| :----- | :--: | :----------- |
| author_id | int | Author's unique ID |
| author_name | string | Author's name |

Table: **CATEGORIES** \
Description: book categories

| Column | Type | Description |
| :----- | :--: | :---------- |
| category_id | int | Category unique ID |
| category_name | string | Category name |

Table: **FORMATS** \
Description: book formats

| Column | Type | Description |
| :----- | :--- | :---------- |
| format_id | int | Format unique ID |
| format_name | string | Format name |

Table: **PLACES** \
Description: book publication places

| Column | Type | Description |
| :----- | :--- | :---------- |
| place_id | int | Publication place unique ID |
| place_name | string | Publication place (city, country) |

Table: **BOOKS** \
Description: books

| Column | Type | Description |
| :----- | :--- | :---------- |
| book_id | long | Book's unique ID |
| isbn10 | string | Book's ISBN-10 |
| isbn13 | long | Book's ISBN-13 |
| title | string | Book's title |
| authors_list | string | List of the book's authors IDs, in the format [##, ##, ##, ...] |
| categories_list | string | List of the book's categories IDs, in the format [##, ##, ##, ...] |
| edition | string | Book edition |
| lang | string | Abbreviation of the book's language (e.g. "en" for English) |
| format_id | int | ID of the book's format |
| publication_place_id | int | ID of the book's publication place |
| publication_date | timestamp | Publication date |
| description | string | Book description |

Table: **BOOKS_AUTHORS** \
Description: many-to-many relationship between books and authors

| Column | Type | Description |
| :----- | :--- | :---------- |
| book_id | long | Book ID |
| author_id | int | Author ID |

Table: **BOOKS_CATEGORIES** \
Description: many-to-many relationship between books and categories

| Column | Type | Description |
| :----- | :--- | :---------- |
| book_id | long | Book ID |
| category_id | int | Category ID |

Table: **AMAZON_REVIEWS** \
Description: customer reviews for books on Amazon

| Column | Type | Description |
| :----- | :--- | :---------- |
| review_id | string | Review unique ID |
| book_id | long | Book ID |
| star_rating | int | The 1-5 star rating of the review |
| helpful_votes | int | Number of helpful votes the review received |
| total_votes | int | Number of total votes the review received |
| verified_purchase | string | Identifies if the review is on a verified purchase (Y or N) |
| customer_id | string | Customer ID in Amazon; can be used to aggregate reviews written by a single author |
| review_date | date | Date when the review was written |
| review_year | int | Year when the review was written |
| review_headline | string | Title of the review |
| review_body | string | Review text |

Table: **DATES** \
Description: metadata on review dates

| Column | Type | Description |
| :----- | :--- | :---------- |
| review_date | date | Review date (yyyy-MM-dd) |
| year | int | Year |
| month | int | Month (numeric) |
| day | int | Day of the month (numeric) |
| week | int | Week number of the year (numeric) |
| weekday | int | Day of the week (numeric) |

# Step 5: Complete Project Write Up

The choice to build a data lake in S3 instead of a data warehouse in Redshift is due mainly to reducing storage costs. Since this new dataset is being put together mainly for analytical purposes, there is no need for the cluster to be online 24/7. The analysis can be done by reading the data lake files. Also, the data being in S3, it becomes availabe to a wider range of users, who can rely on schema-on-read to easily read and wrangle the data.

As the datasets are too big to be wrangled locally by pandas, or even by Spark in standalone mode, it has been chosen to run the ETL pipeline in an EMR cluster containing both Spark and Hadoop applications. Also, since both datasets used by this project are static, the data pipeline was implemented in a Python Jupyter Notebook, in order to be run manually in a one-time execution.

Under other scenarios, the problem should be approached differently. For example:
* In a scenario where the data was increased by 100x, it would be recommended to configure an EMR cluster with more nodes, or choose instance types with higher processing capacity for each node.

* In a real case in which new data would become available in a daily basis, I would orchestrate the workflow setting up a DAG in [Apache AirFlow](https://airflow.apache.org/) splitting the logic included in the notebook `BookReviewsDataLake.ipynb` in Python operators and scheduling the DAG to run once a day.

* If the database had to be accessed by 100+ people, I would load the data into Amazon Redshift, since it is an analytical and distributed database optimized for aggregation and read-heavy workloads. By using a distributed database, it is possible to  improve the replications and partitioning of the tables to get faster query results for each user.